hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/KafkaReadFunction.java

```diff
@@ -67,59 +69,99 @@ public KafkaReadFunction(OperationContext context, String topicName, String boot
         props.put(ConsumerConfig.CLIENT_ID_CONFIG, consumer_id);
 
         props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, IntegerDeserializer.class.getName());
-        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ExternalizableDeserializer.class.getName());
+        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
         props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        
+        // MAX_POLL_RECORDS_CONFIG helped performance in standalone with lower values (default == 500).
+        //  With high values, it spent too much time retrieving records from Kafka.
+        props.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, "100");
+//        props.put(ConsumerConfig.MAX_PARTITION_FETCH_BYTES_CONFIG, "10485760"); // 10% of max == 5242880, default == 1048576
 
-        KafkaConsumer<Integer, Externalizable> consumer = new KafkaConsumer<Integer, Externalizable>(props);
+        KafkaConsumer<Integer, byte[]> consumer = new KafkaConsumer<Integer, byte[]>(props);
         consumer.assign(Arrays.asList(new TopicPartition(topicName, partition)));
 
+        KryoSerialization kryo = new KryoSerialization();
+        kryo.init();
+
         return new Iterator<ExecRow>() {
-            Iterator<ConsumerRecord<Integer, Externalizable>> it = null;
-            
-            Predicate<ConsumerRecords<Integer, Externalizable>> noRecords = records ->
+            Iterator<ConsumerRecord<Integer, byte[]>> it = null;
+            Message prevMessage = new Message();
+
+            Predicate<ConsumerRecords<Integer, byte[]>> noRecords = records ->
                 records == null || records.isEmpty();
             
-            private ConsumerRecords<Integer, Externalizable> kafkaRecords(int maxAttempts) throws TaskKilledException {
+            long totalCount = 0L;
+            int maxRetries = 10;
```

Comment:
What happens when we reach 10 retries? It looks like we silently drop data.

Reply:
Each retry has a 1 minute timeout, so after 10 minutes of retrying, if it hasn't received more data within a batch, it stops polling for data and will `LOG.error(id + " KRF.call Didn't get full batch after " + retries + " retries, got " + totalCount + " records");`.  I've seen data within a batch delayed for several seconds but not as long as minutes, so normally 10 minutes will allow picking up the rest of the data.  In a failure scenario in which the upstream code is unable to send the end of a batch, without the eventual timeout this code will be in an infinite loop tying up a thread / spark task.

Reply:
The problem is we are not failing the operation (as far as I can see). So after 10 minutes we complete the write without error. I think we should have a shorter timeout (1 minute?) and raise an exception.

Reply:
Added exception and shortened timeout in commit https://github.com/splicemachine/spliceengine/pull/4303/commits/063e869b22c07e62976c7e10a8b969397010b898